groups/ntc/ntco/ntco_ioring.cpp

```diff
@@ -0,0 +1,6359 @@
+    // threads driving this proactor. Optionally specify a 'basicAllocator'
+    // used to supply memory. If 'basicAllocator' is 0, the currently
+    // installed default allocator is used.
+    bsl::shared_ptr<ntci::Strand> createStrand(
+        bslma::Allocator* basicAllocator = 0) BSLS_KEYWORD_OVERRIDE;
+
+    // Attach the specified 'socket' to the proactor. Return the
+    // error.
+    ntsa::Error attachSocket(const bsl::shared_ptr<ntci::ProactorSocket>&
+                                 socket) BSLS_KEYWORD_OVERRIDE;
+
+    // Accept the next connection made to the specified 'socket' bound to
+    // the specified 'endpoint'. Return the error.
+    ntsa::Error accept(const bsl::shared_ptr<ntci::ProactorSocket>& socket)
+        BSLS_KEYWORD_OVERRIDE;
+
+    // Connect the specified 'socket' to the specified 'endpoint'. Return
+    // the error.
+    ntsa::Error connect(const bsl::shared_ptr<ntci::ProactorSocket>& socket,
+                        const ntsa::Endpoint& endpoint) BSLS_KEYWORD_OVERRIDE;
+
+    // Enqueue the specified 'data' to the send buffer of the specified
+    // 'socket' according to the specified 'options'. Return the error.
+    // Note that 'data' must not be modified or destroyed until the
+    // operation completes or fails.
+    ntsa::Error send(const bsl::shared_ptr<ntci::ProactorSocket>& socket,
+                     const bdlbb::Blob&                           data,
+                     const ntsa::SendOptions& options) BSLS_KEYWORD_OVERRIDE;
+
+    // Enqueue the specified 'data' to the send buffer of the specified
+    // 'socket' according to the specified 'options'. Return the error.
+    // Note that 'data' must not be modified or destroyed until the
+    // operation completes or fails.
+    ntsa::Error send(const bsl::shared_ptr<ntci::ProactorSocket>& socket,
+                     const ntsa::Data&                            data,
+                     const ntsa::SendOptions& options) BSLS_KEYWORD_OVERRIDE;
+
+    // Dequeue from the receive buffer of the specified 'socket' into the
+    // specified 'data' according to the specified 'options'. Return the
+    // error. Note that 'data' must not be modified or destroyed until the
+    // operation completes or fails.
+    ntsa::Error receive(const bsl::shared_ptr<ntci::ProactorSocket>& socket,
+                        bdlbb::Blob*                                 data,
+                        const ntsa::ReceiveOptions&                  options)
+        BSLS_KEYWORD_OVERRIDE;
+
+    // Shutdown the stream socket in the specified 'direction'. Return the
+    // error.
+    ntsa::Error shutdown(const bsl::shared_ptr<ntci::ProactorSocket>& socket,
+                         ntsa::ShutdownType::Value direction)
+        BSLS_KEYWORD_OVERRIDE;
+
+    // Cancel all outstanding operations initiated for the specified
+    // 'socket'. Return the error.
+    ntsa::Error cancel(const bsl::shared_ptr<ntci::ProactorSocket>& socket)
```

Comment:
It's not a subject of the review, but it looks like that cancellation should be treated as async operation, right? I mean that ideally `ntci::ProactorSocket` would have a method like `processCancelled(...)` to be notified when all operations are cancelled.

Reply:
I think you are correct. I've created a new issue for this work: https://github.com/bloomberg/ntf-core/issues/68.